src/impl_/pymodule.rs

```diff
@@ -71,16 +80,47 @@ impl ModuleDef {
                 ))?;
             }
         }
-        let module = unsafe {
-            Py::<PyModule>::from_owned_ptr_or_err(py, ffi::PyModule_Create(self.ffi_def.get()))?
-        };
-        if self.initialized.swap(true, atomic::Ordering::SeqCst) {
-            return Err(PyImportError::new_err(
-                "PyO3 modules may only be initialized once per interpreter process",
-            ));
+        #[cfg(all(not(PyPy), Py_3_9))]
+        {
+            // If APIs to do so are available, check that the interpreter has not changed.
+            let current_interpreter =
+                unsafe { ffi::PyInterpreterState_GetID(ffi::PyInterpreterState_Get()) };
+            crate::err::error_on_minusone(py, current_interpreter)?;
+            if let Err(initialized_interpreter) = self.interpreter.compare_exchange(
+                -1,
+                current_interpreter,
+                Ordering::SeqCst,
+                Ordering::SeqCst,
+            ) {
+                if initialized_interpreter != current_interpreter {
+                    return Err(PyImportError::new_err(
+                        "PyO3 modules do not yet support subinterpreters, see https://github.com/PyO3/pyo3/issues/576",
+                    ));
+                }
+            }
         }
-        (self.initializer.0)(py, module.as_ref(py))?;
-        Ok(module)
+        #[cfg(all(not(PyPy), not(Py_3_9)))]
```

Comment:
I am a bit confused by the PyPy handling. I would have expected `cfg(any(PyPy, not(Py_3_9)))` here.

As it is written importing on PyPy always succeeds and returns a reference to the existing object. Is this intentional? If so, I would be glad if we could add a code comment to make it explicit.

Reply:
Will do. The idea is that on PyPy there's no such thing as a subinterpreter yet, so no interpreter check is necessary (it would always succeed).

Reply:
Reworked to put the whole check in a `#[cfg(not(PyPy))]` with a comment.